### ISSUE TITLE
Issues with the Android module

### ISSUE THREAD

**Issue by redorav** [premake/premake-core#1256]
I've spent a bit of time trying to create an Android native app build for a small project I have. The way I've set up my premake.lua is as follows:

```lua
workspace("app")

configurations { "Debug", "Release" }
platforms { PlatformMSVC64, PlatformAndroidARM }
cppdialect("c++11")

filter { "platforms:"..PlatformMSVC64 }
    toolset("msc")
    vectorextensions ("sse4.1")
    architecture("x64")

filter { "platforms:"..PlatformAndroidARM }
    system("android")
    architecture("arm")
    vectorextensions("neon")

project ("app_main")
    kind("ConsoleApp")
    files
    {
        srcDir.."/*.cpp",
    }

    includedirs
    {
        srcDir.."/**.h"
    }
	
    filter { "platforms:"..PlatformAndroidARM }
        kind("SharedLib")
        files
        {
            "src/android/android_native_app_glue.h",
            "src/android/android_native_app_glue.c",
        }

project ("app_android")
    kind("AndroidProj")
    androidapplibname("app_main")
    files
    {
        "src/android/AndroidManifest.xml",
        "src/android/build.xml",
        "src/android/project.properties",
        "src/android/res/values/strings.xml",
    }
```

1) The first thing I want to ask what the best place is to write a small tutorial for the Android module, as it's not really as straightforward as the other platform configurations, given the need for androidproj, etc. It's taken me a few hours of trying, restarting, etc to get it working.

2) It seems like androidproj needs to have a reference to the dynamic library where main lives, I know how to do it through the UI but not through premake. I haven't been able to find what api I need to do this. Failure to put the reference produces an error, "Error installing the package. The package ABI 'unknown' is incompatible with the ABI 'armeabiv7a' of device"

3) It also looks like the minute a project with an androidproj file lives alongside other configurations, the other configurations also run the apk process as part of their build. This is clearly unwanted behavior but I don't know how to exclude the androidproj.

Hopefully these are all simple matters that you've had experience with. Thanks!

**Comment by samsinsane** [comment]
The first point seems quite dependent on what happens from #1251. Especially if the AndroidProj is automatically generated from the `*App` project kinds, which seems like it would be significantly more intuitive than the current approach. Anyway, the best place for a tutorial would be in the wiki, you can link to it from [here](https://github.com/premake/premake-core/wiki/Tutorials).

If point 3 is with the sample you've provided, unless I'm mistaken, you'll need `removeplatforms { "PlatformMSVC64" }` in the AndroidProj project.

**Comment by redorav** [comment]
Thanks a lot for pointing me in the right direction, I'll make sure to put a short page there describing how to do it once #1251 is resolved. It's not just how to set up the androidproj but also especially with regards to the build.xml, the manifest, how you need to link the androidproj to the main library, how your main project can't be an *App anymore but needs to be a dynamic lib, etc. If all this is bound to change it's maybe best to wait. Also, the startproject needs to change from being the main executable to being the android project. It seems like currently this can't be done at platform scope which is inconvenient. Perhaps I can make a PR about this.

With regards to number 3, I managed to make it work using removeplatforms("*") and then adding platform(AndroidARM), since I actually have many paltforms I need to remove. Is there an api that does this in one step? Something like restrictplatforms { PlatformA, PlatformB }, if not it's probably easy to set up and I'd be happy to make a PR about it.

**Comment by samsinsane** [comment]
> It seems like currently this can't be done at platform scope which is inconvenient. Perhaps I can make a PR about this.

That wouldn't really make any sense though since it's a workspace setting. Maybe I've misunderstood what you are suggesting?

> With regards to number 3, I managed to make it work using removeplatforms("*") and then adding platform(AndroidARM), since I actually have many paltforms I need to remove. Is there an api that does this in one step? Something like restrictplatforms { PlatformA, PlatformB }, if not it's probably easy to set up and I'd be happy to make a PR about it.

Perhaps more flexible selecting in the `remove*` APIs would be more beneficial than a specific `restrictplatforms` API? I haven't really thought much further than this, perhaps some sort of regex system? It might be complete overkill though.

I vaguely recall there being a discussion about what kind of pattern matching system we should use; PCRE or Lua Patterns, perhaps even something else but I'm not sure where that happened or what the outcome was. @starkos might recall it, but I think it happened 2 or 3 years ago now.

**Comment by starkos** [comment]
There was a discussion about adding a syntax to call directly into the Lua pattern matching; that would definitely still be useful. 

I'm not able to find the discussion now, but we were throwing around ideas like "\`[0-9]\`" as a syntax proposal.

**Comment by redorav** [comment]
@samsinsane Yeah I didn't explain myself too well with regards to startproject, what I meant is to be able to select at _workspace_ scope but filter based on _platform_ or any other criteria, i.e. something like the following:

```lua
workspace("app")

filter { "platforms:"..PlatformMSVC64 }
    startproject("main_app")

filter { "platforms:"..PlatformAndroidARM }
    startproject("main_app_android") -- because this is the apk builder/launcher
```

I don't know whether this is possible to change, because as far as I've read elsewhere the only way to set startproject is to sort them in the sln or something like that? Maybe I'm wrong about this.

As for the other point, the reason I'd want a restrictplatform instead of removeplatform is that the latter sort of implies your current project needs to know about all other platforms, where really it just cares about the ones it needs to work with. I suppose pattern matching is exactly that, where you only add platforms that respond to a certain criteria?

**Comment by samsinsane** [comment]
The startup project is a workspace setting in Visual Studio, it isn't based on a platform/configuration combo. They do have a new way to specify the startup project, which is stored in what appears to be an SQLite database, but I haven't looked into this at all and it could be just as limited.

I would expect the pattern matching system to allow something like `removeplatforms { "![*Android*]" }`. I think there was also an `excludefrombuild` API at one point, which could probably be used with `filter { "system:android" }`, it's a bit more verbose but it would be more intuitive than "removing" a platform.
